<a href="https://colab.research.google.com/github/akmaldika/Task-IF5153-NLP/blob/main/13521070_praktikum3_2a.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Praktikum 3
NIM: 13521070<br>
Nama: Akmal Mahardika Nurwahyu Pratama

## Lab Requirements

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
%pip install datasets tdqm transformers torch

  Preparing metadata (setup.py) ... done
INFO: pip is looking at multiple versions of multiprocess to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 471.6/471.6 kB 16.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 10.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 10.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 15.0 MB/s eta 0:00:00
  Created wheel for tdqm: filename=tdqm-0.0.1-py3-none-any.whl size=1321 sha256=85d5fddad91c7781d50c951b00a4e40594d07f60428cfb487c1201a1e953f325
  Stored in directory: /root/.cache/pip/wheels/37/31/b8/7b711038035720ba0df14376af06e5e76b9bd61759c861ad92
Successfully built tdqm


In [ ]:
class Config:
  BASE_PATH = "/content/drive/MyDrive/ITB/Semester 7/IF5153 Pemrosesan Bahasa Alami/Praktikum3"

Config = Config()

## Dataset

In [ ]:
from datasets import load_dataset

Config.DATASET_NAME = "rotten_tomatoes"

dataset = load_dataset(Config.DATASET_NAME)

dataset

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/7.46k [00:00<?, ?B/s]

train.parquet:   0%|          | 0.00/699k [00:00<?, ?B/s]

validation.parquet:   0%|          | 0.00/90.0k [00:00<?, ?B/s]

test.parquet:   0%|          | 0.00/92.2k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/8530 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/1066 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1066 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 8530
    })
    validation: Dataset({
        features: ['text', 'label'],
        num_rows: 1066
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 1066
    })
})

## Model

In [ ]:
Config.MODEL_NAME = "distilbert/distilbert-base-uncased"
Config.MODEL_PATH = f"{Config.BASE_PATH}/models/{Config.MODEL_NAME}"
Config.TOKENIZER_PATH = f"{Config.BASE_PATH}/models/{Config.MODEL_NAME}_tokenizer"

In [ ]:
import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import os

device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
if os.path.exists(Config.MODEL_PATH) and os.path.exists(Config.TOKENIZER_PATH):
  classification_model = AutoModelForSequenceClassification.from_pretrained(Config.MODEL_PATH, num_labels=2)
  tokenizer = AutoTokenizer.from_pretrained(Config.TOKENIZER_PATH)
else:
  classification_model = AutoModelForSequenceClassification.from_pretrained(Config.MODEL_NAME, num_labels=2)
  tokenizer = AutoTokenizer.from_pretrained(Config.MODEL_NAME)

  classification_model.save_pretrained(Config.MODEL_PATH)
  tokenizer.save_pretrained(Config.TOKENIZER_PATH)

classification_model.to(device)

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert/distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


DistilBertForSequenceClassification(
  (distilbert): DistilBertModel(
    (embeddings): Embeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (transformer): Transformer(
      (layer): ModuleList(
        (0-5): 6 x TransformerBlock(
          (attention): MultiHeadSelfAttention(
            (dropout): Dropout(p=0.1, inplace=False)
            (q_lin): Linear(in_features=768, out_features=768, bias=True)
            (k_lin): Linear(in_features=768, out_features=768, bias=True)
            (v_lin): Linear(in_features=768, out_features=768, bias=True)
            (out_lin): Linear(in_features=768, out_features=768, bias=True)
          )
          (sa_layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (ffn): FFN(
            (dropout): Dropout(p=0.1, inplace=False)
 

## Train

In [ ]:
Config.TOKENIZED_DATASET_PATH = f"{Config.BASE_PATH}/tokenized_datasets"

In [ ]:
from transformers import BatchEncoding, DataCollatorWithPadding
from datasets import Dataset, load_from_disk

def tokenize_function(isinstances_text: Dataset) -> BatchEncoding:
    return tokenizer(isinstances_text["text"],
                     padding="max_length",
                     truncation=True)

if os.path.exists(Config.TOKENIZED_DATASET_PATH):
  tokenized_datasets = load_from_disk(Config.TOKENIZED_DATASET_PATH)
else:
  tokenized_datasets = dataset.map(tokenize_function)
  tokenized_datasets = tokenized_datasets.rename_column("label", "labels")
  tokenized_datasets.set_format(type="torch", columns=["input_ids", "attention_mask", "labels"])
  tokenized_datasets.save_to_disk(Config.TOKENIZED_DATASET_PATH)

data_collator = DataCollatorWithPadding(tokenizer=tokenizer)
display(data_collator)
display(tokenized_datasets)

Map:   0%|          | 0/8530 [00:00<?, ? examples/s]

Map:   0%|          | 0/1066 [00:00<?, ? examples/s]

Map:   0%|          | 0/1066 [00:00<?, ? examples/s]

Saving the dataset (0/1 shards):   0%|          | 0/8530 [00:00<?, ? examples/s]

Saving the dataset (0/1 shards):   0%|          | 0/1066 [00:00<?, ? examples/s]

Saving the dataset (0/1 shards):   0%|          | 0/1066 [00:00<?, ? examples/s]

DataCollatorWithPadding(tokenizer=DistilBertTokenizerFast(name_or_path='distilbert/distilbert-base-uncased', vocab_size=30522, model_max_length=512, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("[PAD]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	100: AddedToken("[UNK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	101: AddedToken("[CLS]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	102: AddedToken("[SEP]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	103: AddedToken("[MASK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}, padding=True, max_length=None, pad_to_multiple_of=None, return_tensor

DatasetDict({
    train: Dataset({
        features: ['text', 'labels', 'input_ids', 'attention_mask'],
        num_rows: 8530
    })
    validation: Dataset({
        features: ['text', 'labels', 'input_ids', 'attention_mask'],
        num_rows: 1066
    })
    test: Dataset({
        features: ['text', 'labels', 'input_ids', 'attention_mask'],
        num_rows: 1066
    })
})

In [ ]:
Config.BATCH_SIZE = 8
Config.NUM_EPOCHS = 1
Config.TUNNED_PATH = f"{Config.BASE_PATH}/tuned_models/{Config.MODEL_NAME}"
Config.RESULT_PATH = f"{Config.BASE_PATH}/results"

In [ ]:
from transformers import Trainer, TrainingArguments
from sklearn.metrics import accuracy_score

def compute_metrics(pred):
  labels = pred.label_ids
  preds = pred.predictions.argmax(-1)
  accuracy = accuracy_score(labels, preds)
  return {'accuracy': accuracy}

if os.path.exists(Config.TUNNED_PATH):
  trainer = Trainer.from_pretrained(Config.TUNNED_PATH)
else:
  training_args = TrainingArguments(
      output_dir=Config.RESULT_PATH,
      num_train_epochs=Config.NUM_EPOCHS,
      per_device_train_batch_size=Config.BATCH_SIZE,
      per_device_eval_batch_size=Config.BATCH_SIZE,

      eval_strategy="epoch",
      fp16=True,
      )

  trainer = Trainer(
      model=classification_model,
      args=training_args,
      train_dataset=tokenized_datasets["train"],
      eval_dataset=tokenized_datasets["validation"],
      compute_metrics=compute_metrics
      )

  trainer.train()
  trainer.save_model(Config.TUNNED_PATH)

/usr/local/lib/python3.10/dist-packages/accelerate/accelerator.py:494: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  self.scaler = torch.cuda.amp.GradScaler(**kwargs)


## Validation

In [ ]:
# Evaluate on validation set
trainer.evaluate(tokenized_datasets["validation"])

{'eval_loss': 0.3566869795322418,
 'eval_accuracy': 0.8536585365853658,
 'eval_runtime': 7.1666,
 'eval_samples_per_second': 148.745,
 'eval_steps_per_second': 18.698,
 'epoch': 1.0}

## Test

### Test from test data

In [ ]:
from sklearn.metrics import classification_report
# Predict on test set
predictions = trainer.predict(tokenized_datasets["test"])

pradictions_labels = predictions.predictions.argmax(-1)

print(classification_report(predictions.label_ids, pradictions_labels))

              precision    recall  f1-score   support

           0       0.84      0.86      0.85       533
           1       0.86      0.84      0.85       533

    accuracy                           0.85      1066
   macro avg       0.85      0.85      0.85      1066
weighted avg       0.85      0.85      0.85      1066



### Test from sintetic data

In [ ]:
import torch

text = [
    "This is a positive review.",
    "This is a negative review.",
    "This is a neutral review.",
    "I love this movie",
    "I hate this movie",
    "This movie is ok",
]

inputs = tokenizer(text, padding=True, truncation=True, return_tensors="pt")

with torch.no_grad():
    inputs = {k: v.to(classification_model.device) for k, v in inputs.items()}
    outputs = classification_model(**inputs)
    display("Class logits. Shape:", outputs.logits.shape)
    display(outputs.logits)
    display("Class predictions:")
    classes = outputs.logits.argmax(dim=1)
    display("Classes", classes.cpu().detach().numpy().tolist())



display(predictions)


'Class logits. Shape:'

torch.Size([6, 2])

tensor([[-1.5557,  1.8682],
        [ 1.7637, -1.5947],
        [ 0.7344, -0.6416],
        [-1.4912,  1.6826],
        [ 1.5879, -1.3750],
        [-1.0322,  1.0957]], device='cuda:0')

'Class predictions:'

'Classes'

[1, 0, 0, 1, 0, 1]

PredictionOutput(predictions=array([[-1.0527344 ,  1.4267578 ],
       [-1.8037109 ,  2.2851562 ],
       [ 0.96728516, -0.87646484],
       ...,
       [ 1.5205078 , -1.3886719 ],
       [ 1.5205078 , -1.4355469 ],
       [ 2.0410156 , -1.8173828 ]], dtype=float32), label_ids=array([1, 1, 1, ..., 0, 0, 0]), metrics={'test_loss': 0.37414246797561646, 'test_accuracy': 0.849906191369606, 'test_runtime': 6.4047, 'test_samples_per_second': 166.441, 'test_steps_per_second': 20.922})

## Analisis

Secara umum model hasil tunning:
- model memiliki akurasi 0.85 terhadap data test
- model memiliki *precision* dan *recall* yang seimbang.
- data sintetis memiliki akurasi yang bagus (peneliti menganggap data neutral disana adalah negatif)